INT-4444: Introduce `@Reactive` & `reactive()`

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4444

Right now the high-level API creates a `ReactiveStreamsConsumer`
only when the input channel is a `Publisher<?>` impl or target handler
is a `ReactiveMessageHandler`

* Add `@Reactive[] reactive()` attribute to messaging annotations
* Add `ConsumerEndpointSpec.reactive()`
Both options point to the same `ConsumerEndpointFactoryBean.setReactiveCustomizer()`
making the target endpoint always as a `ReactiveStreamsConsumer` independently of
the input channel and target handler
* Use the `Function` to customize a source `Flux` from the channel
* Test and document a new feature

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
INT-4444: Introduce `@Reactive` & `reactive()`

JIRA: https://jira.spring.io/browse/INT-4444

Right now the high-level API creates a `ReactiveStreamsConsumer`
only when the input channel is a `Publisher<?>` impl or target handler
is a `ReactiveMessageHandler`

* Add `@Reactive[] reactive()` attribute to messaging annotations
* Add `ConsumerEndpointSpec.reactive()`
Both options point to the same `ConsumerEndpointFactoryBean.setReactiveCustomizer()`
making the target endpoint always as a `ReactiveStreamsConsumer` independently of
the input channel and target handler
* Use the `Function` to customize a source `Flux` from the channel
* Test and document a new feature

artembilan committed:
* Fix links in docs

artembilan committed:
* Fix `ReactiveStreamsTests`

garyrussell commented on src/reference/asciidoc/dsl.adoc:
```diff
@@ -262,6 +262,31 @@ See https://docs.spring.io/spring-integration/api/org/springframework/integratio
+[[java-dsl-reactive]]
+=== The `reactive()` Endpoint
+
+Starting with version 5.5, the `ConsumerEndpointSpec` provide a `reactive()` configuration property with an optional customizer `Function<? super Flux<Message<?>>, ? extends Publisher<Message<?>>>`.
+This option makes the target endpoint as a `ReactiveStreamsConsumer` instance independently of the input channel type, which is turned to the `Flux` via `IntegrationReactiveUtils.messageChannelToFlux()`.
+The provided function is used from the `Flux.transform()` operator to customize (`publishOn()`, `log()`, `doOnNext()` etc.) a reactive stream source from the input channel.
```
```suggestion
Starting with version 5.5, the `ConsumerEndpointSpec` provides a `reactive()` configuration property with an optional customizer `Function<? super Flux<Message<?>>, ? extends Publisher<Message<?>>>`.
This option configures the target endpoint as a `ReactiveStreamsConsumer` instance, independently of the input channel type, which is converted to a `Flux` via `IntegrationReactiveUtils.messageChannelToFlux()`.
The provided function is used from the `Flux.transform()` operator to customize (`publishOn()`, `log()`, `doOnNext()` etc.) a reactive stream source from the input channel.
```

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/annotation/Aggregator.java:
```diff
@@ -104,8 +104,16 @@
+	 * Only one {@link Reactive} element is allowed.
+	 * Mutually exclusive with {@link #poller()}.
+	 */
+	Reactive[] reactive() default { };
```
OK. `@Reactive("not_provided")` (or similar marker) will work. Otherwise I would argue how to determine "no reactive" then: the function is optional, but users really would like to make an endpoint reactive independently of the input channel.

garyrussell replied:
I actually don't have a strong opinion on this, given that we do the same with `@Poller` - so feel free to reject the suggestion.

garyrussell replied:
(all `@Poller` attributes have defaults too).

artembilan replied:
Yeah... I think at the moment we introduced a `@Poller` we just didn't think about some some `not_provided` marker. 
I believe we can break it and bring it back to a single annotation in `6.0`. See `ValueConstants.DEFAULT_NONE`

artembilan commented:
Forgot to mention: let's consider an XML tag as possible external contribution. Too much places to take care about:

<img width="428" alt="Poller_in_XML" src="https://user-images.githubusercontent.com/1071590/110177634-4b4f6d00-7dd3-11eb-9c49-965b51f52ab5.PNG">

Plus it is not so natural to declare `Function` beans in the XML config.
It is a bit awkward with that `@Reactive` annotation, but there is no other choice with them unless bean references...

The Java DSL is the best choice for this type of configuration! 😄

artembilan committed:
* Rework `reactive()` attribute of messaging annotations ot a single `@Reactive` value
with default as `@Reactive(ValueConstants.DEFAULT_NONE)`
* Fix language in docs
* Fix `MessagingAnnotationUtils.resolveAttribute()` to use `requiredType.isInstance()`
instead of comparing classes since annotation instances are `Proxy` at runtime
